File: 1StudentInfo.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
// Store data records in a linked list.

// Structure of Student record 
struct student{
    int roll;
    char *name;
    char *phn;
    };
typedef struct student stud; // typdef to stud for ease to use.


// Structure of linear linked list
struct node{
    stud data;
    struct node *next;
};
struct node *head=NULL; // Assigning head to null

// This function will create a new node. It is used by create function.
struct node *createNode(stud data){
    struct node *new_node=(struct node*)malloc(sizeof(struct node)); //creating space in heap
    new_node->data=data;  //assigning data and setting next pointer as null
    new_node->next=NULL; 
    return new_node;
}

// This function will create a new node and at at the end O(n) time.
void create(stud data){
    struct node *new_node=createNode(data);
    if(head==NULL){
        head=new_node; // if first element to insert
    }
    else{
        struct node *tmp=head;
        while(tmp->next!=NULL){
            tmp=tmp->next;
        }
        tmp->next=new_node; // inserting at the end
    }
}

// this function will traverse entire list and print data sequentially
void read(){
    if(head==NULL){
        printf("Empty list"); //undeflow condition
        return;
    }
    printf("\nreading data\n");
    struct node *tmp=head;
    while(tmp!=NULL){
        printf("\nRoll: %d\n",tmp->data.roll);
        printf("name: %s\n",tmp->data.name);
        printf("phn: %s\n",tmp->data.phn);

        tmp=tmp->next;
    }


}

// this function will update the data using roll number as key
void update(int roll,stud data){
    if(head==NULL){
        printf("Empty list"); //undeflow condition
        return;
    }
    struct node *tmp=head;
    // traversing untill find a match
    while(tmp->data.roll!=roll){ 
        tmp=tmp->next;
    }
    tmp->data=data; // update the record


}
// This function will delete the record based on roll number
void delete(int roll){
     if(head==NULL){
        printf("Empty list"); //undeflow condition
        return;
    }
     struct node *tmp=head;
     // if first record is found match delete it
     if(tmp->data.roll==roll){
         head=head->next; 
         free(tmp);
     }
     else{// else traverse the entire list, delete once found match
         while(tmp->next->data.roll!=roll){
             tmp=tmp->next;
         }
        struct node *current=tmp->next;
        tmp->next=current->next;
        free(current);
     }

}

// Main program 

int main(){
   
    stud data1={1,"Abhi","1234"};
    stud data2={2,"Amit","4321"};
    stud data3={3,"Ajay","3421"};
    stud data4={4,"Shyam","23456"};

    create(data1);
    create(data2);
    create(data3);
    create(data4);

    read();
    data1.name="Abhishek"; // updating the name of roll no  1
    update(1,data1);

    read();

    delete(2); // deleting the record with roll number 2

    read();

}

File: 1StudentInfo.py
## Store data records in a linked list.

## Template of linear linked list
class createNode:
    def __init__(self,data):
        self.data=data
        self.next=None
        
## this class will contain all the methods(CRUD)
class Student_Details:
    def __init__(self):
        self.head=None ## initialising head as none
      
    ## create a record and add at the end
    def create(self,data):
        new_node=createNode(data)
        if(self.head==None):
            self.head=new_node ## if first element 
        else:
            tmp=self.head
            while(tmp.next!=None):
                tmp=tmp.next 
            tmp.next=new_node ## adding the record at end
        
    def read(self):
        if(self.head==None):
            print("Empty List") ##underflow condition
            return
        
        print("\nReading data")
        tmp=self.head
        while(tmp!=None):
            print(tmp.data) ## printing data sequentially
            tmp=tmp.next
       
    ## this method will update the student data based on Roll Number(unique)
    def update(self,roll,data):
        
        if(self.head==None):
            print("Empty List") ##underflow condition
            return
        
        tmp=self.head
        while(tmp.data["roll"]!=roll): 
            tmp=tmp.next
        tmp.data=data
     
    ## this method will delete the student data based on Roll Number(unique)
    def delete(self,roll):
        
        if(self.head==None):
            print("Empty List") ##underflow condition
            return
        
        tmp=self.head
        ## if the first record is itself matched thn  delete  
        if(tmp.data["roll"]==roll):
            self.head=self.head.next
            del(tmp)
        ## if not the first record thn match rest of the node and delete 
        else:
            while(tmp.next.data["roll"]!=roll):
                tmp=tmp.next
            current=tmp.next
            tmp.next=current.next
            del(current)
    
if __name__=="__main__":
   
 ## data is stored in Json format
    data1={"roll":1,
           "name":"Abhi",
           "phn":"1234"}
    data2={"roll":2,
           "name":"Amit",
           "phn":"4321"}
    data3={"roll":3,
           "name":"Ajay",
           "phn":"3421"}
    data4={"roll":4,
           "name":"Shyam",
           "phn":"23456"}
    
    ## Initiating stud object and storing data
    Stud=Student_Details()
    Stud.create(data1)
    Stud.create(data2)
    Stud.create(data3)
    Stud.create(data4)
    
    Stud.read()
    ## udating roll number 1 name
    data1["name"]="Abhishek"
    roll=1
    Stud.update(roll,data1)
    
    Stud.read()
    ## deleting student with roll 2
    Stud.delete(roll=2)
    
    Stud.read()
